### PR TITLE
Python 3 Upgrade

### PR DESCRIPTION
**Pull request by fanpu**
### Plan:
Unfortunately, we have almost no test coverage so regressions must be detected by hand. Our goal is to port Tango to Python 3 while still maintaining Python 2 compatibility.

1. Follow Futurize porting guide (http://python-future.org/automatic_conversion.html) with manual inspection. For each stage, create PRs that merge into this PR so it is clearer which changes are being introduced at which stage for accountability. 
2. Test Tango CLI on Python 3 manually (maybe write some tests?)
3. Test Tango on Python 3 with Autolab
4. Test Tango CLI on Python 2 manually
5. Test Tango on Python 2 with Autolab

Tested with all CLI commands given in https://autolab.github.io/docs/tango-cli/

This is now ready for review!

### CODE REVIEW

**Comment by fanpu** [comment]
I have reviewed the stage 1 fixes and they LGTM

File: jobManager.py
Comment (victorhuangwq) [inline jobManager.py:78]: this fix does seem to address the issue mentioned here https://github.com/autolab/Tango/issues/138.

File: Dockerfile
Comment (victorhuangwq) [inline Dockerfile:2]: there will be issues when obtaining output from autograder if changed to 18.04 based on https://github.com/autolab/Tango/issues/153

there are relevant patch in there to address it. Has that been added?

File: jobQueue.py
Comment (victorhuangwq) [inline jobQueue.py:140]: Is self.liveJobs a dictionary?

In that case it would be more suitable to change it to `if str(id) in self.liveJobs` (w/o calling keys) for python 3
instead of converting it into a list first.

https://stackoverflow.com/questions/1602934/check-if-a-given-key-already-exists-in-a-dictionary

File: preallocator.py
Comment (victorhuangwq) [inline preallocator.py:34]: same comment with regards to the key situation in jobqueue

File: tangoObjects.py
Comment (victorhuangwq) [inline tangoObjects.py:327]: same comment as above

File: vmms/distDocker.py
Comment (victorhuangwq) [inline vmms/distDocker.py:8]: As you have changed the default docker image to xenial the issue here would happen https://github.com/autolab/Tango/issues/138, 

do update with the patch from here
https://github.com/mtoupsUNO/Tango/commit/b97bc1e1eaa000644d4e030e44e1c96341c2d120

File: jobQueue.py
Comment (fanpu) [reply jobQueue.py:140]: It's either a TangoNativeDictionary or TangoRemoteDictionary (which exports the same interface) which is not iterable so the `x in y` syntax would not work here

File: preallocator.py
Comment (fanpu) [reply preallocator.py:34]: self.machines here is also either a TangoNativeDictionary or TangoRemoteDictionary

File: tangoObjects.py
Comment (fanpu) [reply tangoObjects.py:327]: Works here, this is using python dict

File: vmms/distDocker.py
Comment (fanpu) [reply vmms/distDocker.py:8]: Patch applied

File: Dockerfile
Comment (fanpu) [reply Dockerfile]: xenial is actually 16.04, but I just changed it to 18.04. This Dockerfile is for the Tango Dockerfile, I did not modify the autograding Dockerfile in this PR which is still on 14.04 (should fix that soon though)

**Comment by fanpu** [comment]
@victorhuangwq Thanks for your comments! I've rectified the issues you raised, do look at it again when you're available :D

File: tangoObjects.py
Comment (victorhuangwq) [inline tangoObjects.py:332]: doesn't it return a list here? 

Wouldn't calling list again on keys() be redundant in the other lines I have highlighted above? Such as in
`if vmName not in list(self.machines.keys()):`

File: tangoObjects.py
Comment (victorhuangwq) [inline tangoObjects.py:296]: It seems to be returning a list when keys are called on the TangoRemoteDictionary.

File: preallocator.py
Comment (victorhuangwq) [reply preallocator.py:34]: refer to comment below

**Comment by cg2v** [comment]
This should not block the upgrade, since I believe the current code to be correct, but most if not all of the keys() calls in the code do not require the list form of the keys (you only need to listify the keys if you either want to use [] to access elements of the key list, or if you will alter the dict while iterating over the keys). Furthermore, a more readable way to do
`if vmName in self.machines.keys():`

would be to implement __contains__ in the Tango*Dictionary classes themselves and then call

`if vmName in self.machines:`

instead. TangoRemoteDictionary could use HEXISTS to implement __contains__ and not have to fetch everything.

File: jobManager.py
Comment (cg2v) [inline jobManager.py:132]: Another unnecessary list()

File: preallocator.py
Comment (cg2v) [inline preallocator.py:49]: another unnecessary list()

**Comment by fanpu** [comment]
> This should not block the upgrade, since I believe the current code to be correct, but most if not all of the keys() calls in the code do not require the list form of the keys (you only need to listify the keys if you either want to use [] to access elements of the key list, or if you will alter the dict while iterating over the keys). Furthermore, a more readable way to do
> `if vmName in self.machines.keys():`
> 
> would be to implement `__contains__` in the Tango*Dictionary classes themselves and then call
> 
> `if vmName in self.machines:`
> 
> instead. TangoRemoteDictionary could use HEXISTS to implement `__contains__` and not have to fetch everything.

@cg2v  This is a great suggestion, I will add it as a future enhancement instead of adding that to this PR to avoid making this PR bigger than it already is.

Thank you and really appreciate the spotting of other redundant wrappers of `list` as well!

**Review by victorhuangwq — APPROVED** [review]
**Testing**: Tested Tango3 branch by grading various assignments through Autolab, and it works as expected.

I have verified that:
1. Patch from https://github.com/autolab/Tango/issues/153 has been applied
2. Unnecessary lists, as mentioned by both @cg2v  and I, has been removed
3. Dockerfile is now back on 18.04 (change removed)

This pull request LGTM :)

**Comment by victorhuangwq** [comment]
@fanpu Before you merge in however, please create an alternate Python 2 branch, preferably called `master-python-2` to hold the old code. 

At the same time, here are the list of things we should also do alongside this PR. I can help with it as well. 

- [ ] Update the description on the readme under the section **Python 3 Upgrade.** I do believe the wording that was used in the Autolab readme regarding rails 4 support is good. 
- [ ] Add a new release, marking this as a new version of Tango. (v2.0.0 maybe?)

**Pull request merged**
